throttle/contract_tx_limiter_middleware.go

```diff
@@ -0,0 +1,123 @@
+	"github.com/loomnetwork/loomchain"
+	"github.com/loomnetwork/loomchain/auth"
+	udw "github.com/loomnetwork/loomchain/builtin/plugins/user_deployer_whitelist"
+	"github.com/loomnetwork/loomchain/vm"
+	"github.com/pkg/errors"
+)
+
+const refreshInterval int64 = 15 * 60
+
+type contractTxLimiter struct {
+	// contract_address to limiting parametres structure
+	contractToTierMap map[string]udw.Tier
+	// track of no. of txns in previous blocks per contract
+	contractToBlockTrx map[string]map[int64]int64
+	// to refresh the contractToTierMap when configuration changes
+	lastUpdated int64
+}
+
+var TxLimiter *contractTxLimiter
+
+func (txl *contractTxLimiter) isAccountLimitReached(contractAddr loom.Address, curBlockHeight int64) bool {
+	tier, ok := txl.contractToTierMap[contractAddr.String()]
+	if !ok {
+		return false
+	}
+	blockTxns, ok := txl.contractToBlockTrx[contractAddr.String()]
+	if !ok {
+		return false
+	}
+	minBlockHeight := curBlockHeight - int64(tier.BlockRange) + 1
+	sum := int64(0)
+	for blockHeight, txns := range blockTxns {
+		if blockHeight < minBlockHeight {
+			delete(blockTxns, blockHeight)
+		} else {
+			sum = sum + txns
+		}
+	}
+	if sum > int64(tier.MaxTx) {
+		return false
+	}
+	return true
+}
+
+func (txl *contractTxLimiter) updateState(contractAddr loom.Address, curBlockHeight int64) {
+	_, ok := TxLimiter.contractToBlockTrx[contractAddr.String()]
+	if !ok {
+		TxLimiter.contractToBlockTrx[contractAddr.String()] = make(map[int64]int64, 0)
+	}
+	_, ok = TxLimiter.contractToBlockTrx[contractAddr.String()][curBlockHeight]
+	if !ok {
+		TxLimiter.contractToBlockTrx[contractAddr.String()][curBlockHeight] = 1
+	} else {
+		TxLimiter.contractToBlockTrx[contractAddr.String()][curBlockHeight]++
+	}
+}
+
+// NewContractTxLimiterMiddleware add another tx limiter that limits how many CallTx(s) can be sent to an EVM contract within a pre-configured block range
+func NewContractTxLimiterMiddleware(
+	createUserDeployerWhitelistCtx func(state loomchain.State) (contractpb.Context, error),
+) loomchain.TxMiddlewareFunc {
+	TxLimiter = &contractTxLimiter{
+		contractToBlockTrx: make(map[string]map[int64]int64, 0),
+		lastUpdated:        time.Now().Unix() - refreshInterval, // updation time is fifteen minutes
+	}
+	return loomchain.TxMiddlewareFunc(func(
+		state loomchain.State,
+		txBytes []byte,
+		next loomchain.TxHandlerFunc,
+		isCheckTx bool,
+	) (res loomchain.TxHandlerResult, err error) {
+		if !isCheckTx {
+			return next(state, txBytes, isCheckTx)
+		}
+		var nonceTx auth.NonceTx
+		if err := proto.Unmarshal(txBytes, &nonceTx); err != nil {
+			return res, errors.Wrap(err, "throttle: unwrap nonce Tx")
+		}
+		var tx loomchain.Transaction
+		if err := proto.Unmarshal(nonceTx.Inner, &tx); err != nil {
+			return res, errors.New("throttle: unmarshal tx")
+		}
+		if tx.Id != callId {
+			return next(state, txBytes, isCheckTx)
+		}
+		// updated only if tx.Id == callId
+		if TxLimiter.lastUpdated+refreshInterval <= time.Now().Unix() {
```

Comment:
Need to have two condition here.

`TxLimiter.contractToTierMap == nil || TxLimiter.lastUpdated+refreshInterval <= time.Now().Unix()`

Reply:
Also, let's execute this code only in case tx is evm tx.

Reply:
done